autometa/common/external/samtools.py

```diff
@@ -51,24 +51,29 @@ def sort(sam, bam, nproc=mp.cpu_count()):
     
     """
 
-    cmd = f'samtools view -@{nproc} -bS {sam} | samtools sort -@{nproc} -o {bam}'
+    samtools_out_dir = os.path.dirname(os.path.abspath(bam))
+    outfile_tail = os.path.basename(bam)
+    tempdir = tempfile.mkdtemp(suffix = None, prefix = 'samtools', dir = samtools_out_dir)
+    samtools_outfile_tail = os.path.join(samtools_out_dir, outfile_tail)
+    temp_outfile_tail = os.path.join(tempdir, outfile_tail)
+    cmd = f'samtools view -@{nproc} -bS {sam} | samtools sort -@{nproc} -o {temp_outfile_tail}'
     logger.debug(f'cmd: {cmd}')
-    samtools_out_dir = os.path.dirname(bam)
-    tempdir = tempfile.mkdtemp(suffix=None, prefix='samtools', dir=samtools_out_dir)
-    os.chdir(tempdir)
     samtools_err = os.path.join(samtools_out_dir, 'samtools.err')
     samtools_out = os.path.join(samtools_out_dir, 'samtools.out')
     with open(samtools_out, 'w') as stdout, open(samtools_err, 'w') as stderr:
         try:
             retcode = subprocess.call(cmd, stdout=stdout, stderr=stderr, shell=True)
             if retcode != 0:
                 raise ChildProcessError(f'Samtools not successfully run, retcode: {retcode}')
```

Comment:
You could shorten this statement to say "sort failed" rather than "not successfully run". (Minor edit to help screen wrapping). This is more descriptive, to the function in which the child process is being raised and more concise. When the exception is raised, the traceback will need to the filename hence "samtools" will already be referenced.